deep_qa/layers/attention/weighted_sum.py

```diff
@@ -71,39 +71,51 @@ def compute_output_shape(self, input_shapes):
 
     @overrides
     def call(self, inputs, mask=None):
+        # pylint: disable=redefined-variable-type
         matrix, attention_vector = inputs
-        matrix_shape = K.int_shape(matrix)
-        matrix = self._expand_matrix_if_necessary(matrix, matrix_shape[:-1], attention_vector)
+        matrix = self._expand_matrix_if_necessary(matrix, attention_vector, K.ndim(matrix) - 1)
         if mask is None:
             matrix_mask = None
         else:
             matrix_mask = mask[0]
         if self.use_masking and matrix_mask is not None:
-            matrix_mask = self._expand_matrix_if_necessary(matrix_mask, matrix_shape[:-1], attention_vector)
+            matrix_mask = self._expand_matrix_if_necessary(matrix_mask, attention_vector, K.ndim(matrix) - 1)
             # Doing a multiply here instead of a `switch` to avoid allocating another large tensor.
             matrix = K.cast(K.expand_dims(matrix_mask), 'float32') * matrix
         return K.sum(K.expand_dims(attention_vector, axis=-1) * matrix, -2)
 
     @staticmethod
-    def _expand_matrix_if_necessary(matrix, matrix_shape, attention_vector):
+    def _expand_matrix_if_necessary(matrix, attention_vector, num_matrix_dims: int):
```

Comment:
You mean do `K.ndim(matrix) - 1`?  The trouble is that `K.ndim(matrix)` and `K.ndim(mask)` are different, so I need some way to distinguish them inside the function.  Even though it says `matrix`, something this method will be passed the `mask`.